File: apps/backend/src/routes/account.py
import sqlite3

from flask import Blueprint, jsonify, request
from src.db.db_manager import db_manager
from src.services.cookie_service import get_cookie_service

# 创建蓝图
bp = Blueprint("account", __name__)


@bp.route("/getAccounts", methods=["GET"])
def getAccounts():
    """快速获取所有账号信息，不进行cookie验证"""
    try:
        with sqlite3.connect(db_manager.get_db_path()) as conn:
            conn.row_factory = sqlite3.Row
            cursor = conn.cursor()
            cursor.execute("""
            SELECT * FROM user_info""")
            rows = cursor.fetchall()
            rows_list = [list(row) for row in rows]

            print("\n📋 当前数据表内容（快速获取）：")
            for row in rows:
                print(row)

            return jsonify({"code": 200, "msg": None, "data": rows_list}), 200
    except Exception as e:
        print(f"获取账号列表时出错: {str(e)}")
        return jsonify({"code": 500, "msg": f"获取账号列表失败: {str(e)}", "data": None}), 500


@bp.route("/getValidAccounts", methods=["GET"])
async def getValidAccounts():
    # 获取账号ID参数，如果提供则只验证单个账号
    account_id = request.args.get("id")
    # 强制刷新参数，跳过冷却期检查
    force_refresh = request.args.get("force", "false").lower() == "true"

    # 验证冷却期（秒）- 刚登录成功的账号在此时间内跳过二次验证
    VALIDATION_COOLDOWN_SECONDS = 300  # 5分钟冷却期，减少不必要的浏览器验证

    with sqlite3.connect(db_manager.get_db_path()) as conn:
        cursor = conn.cursor()

        # 根据是否提供id参数决定查询范围
        if account_id:
            cursor.execute(
                """
            SELECT id, type, filePath, userName, status, group_id, created_at, last_validated_at
            FROM user_info WHERE id = ?""",
                (account_id,),
            )
        else:
            cursor.execute("""
            SELECT id, type, filePath, userName, status, group_id, created_at, last_validated_at
            FROM user_info""")

        rows = cursor.fetchall()
        rows_list = [list(row) for row in rows]

        print("\n📋 当前数据表内容：")
        for row in rows:
            print(row)

        for row in rows_list:
            account_id_val = row[0]
            account_type = row[1]
            file_path = row[2]
            current_status = row[4]
            last_validated_at = row[7]  # 最后验证时间

            # 检查是否在验证冷却期内（除非强制刷新）
            should_skip_validation = False
            if not force_refresh and last_validated_at:
                from datetime import datetime, timedelta

                try:
                    # SQLite 时间格式
                    validated_time = datetime.strptime(last_validated_at, "%Y-%m-%d %H:%M:%S")
                    cooldown_end = validated_time + timedelta(seconds=VALIDATION_COOLDOWN_SECONDS)
                    if datetime.now() < cooldown_end:
                        should_skip_validation = True
                        print(
                            f"⏭️ 账号 {row[3]} 在冷却期内，跳过验证 (上次验证: {last_validated_at})"
                        )
                except Exception as e:
                    print(f"⚠️ 解析验证时间失败: {e}")

            if should_skip_validation:
                # 保持当前状态，不重新验证
                continue

            flag = await get_cookie_service().check_cookie(account_type, file_path)
            if not flag:
                row[4] = 0
                cursor.execute(
                    """
                UPDATE user_info
                SET status = ?, last_validated_at = CURRENT_TIMESTAMP
                WHERE id = ?
                """,
                    (0, account_id_val),
                )
                conn.commit()
                print("✅ 用户状态已更新")
            else:
                # 如果状态正常，确保数据库中也是正常状态
                row[4] = 1
                cursor.execute(
                    """
                UPDATE user_info
                SET status = ?, last_validated_at = CURRENT_TIMESTAMP
                WHERE id = ?
                """,
                    (1, account_id_val),
                )
                conn.commit()
                if current_status != 1:
                    print("✅ 用户状态已更新为正常")

        for row in rows:
            print(row)

        return jsonify({"code": 200, "msg": None, "data": rows_list}), 200


@bp.route("/getAccountStatus", methods=["GET"])
async def getAccountStatus():
    """获取单个账号的实时状态"""
    account_id = request.args.get("id")

    if not account_id:
        return jsonify({"code": 400, "msg": "缺少账号ID参数", "data": None}), 400

    try:
        with sqlite3.connect(db_manager.get_db_path()) as conn:
            cursor = conn.cursor()
            # 查询账号信息
            cursor.execute("SELECT * FROM user_info WHERE id = ?", (account_id,))
            row = cursor.fetchone()

            if not row:
                return jsonify({"code": 404, "msg": "账号不存在", "data": None}), 404

            row_list = list(row)
            # 验证cookie状态
            flag = await get_cookie_service().check_cookie(row_list[1], row_list[2])
            current_status = 1 if flag else 0

            # 更新数据库状态
            if row_list[4] != current_status:
                cursor.execute(
                    """
                UPDATE user_info
                SET status = ?
                WHERE id = ?
                """,
                    (current_status, account_id),
                )
                conn.commit()
                row_list[4] = current_status

            # 返回标准化的状态信息
            return (
                jsonify(
                    {
                        "code": 200,
                        "msg": None,
                        "data": {
                            "id": row_list[0],
                            "type": row_list[1],
                            "filePath": row_list[2],
                            "userName": row_list[3],
                            "status": current_status,
                            "statusText": "正常" if current_status == 1 else "异常",
                            "isValid": flag,
                        },
                    }
                ),
                200,
            )

    except Exception as e:
        print(f"获取账号状态时出错: {str(e)}")
        return jsonify({"code": 500, "msg": f"获取账号状态失败: {str(e)}", "data": None}), 500


@bp.route("/deleteAccount", methods=["GET"])
def delete_account():
    account_id = int(request.args.get("id"))

    try:
        # 获取数据库连接
        with sqlite3.connect(db_manager.get_db_path()) as conn:
            conn.row_factory = sqlite3.Row
            cursor = conn.cursor()

            # 查询要删除的记录
            cursor.execute("SELECT * FROM user_info WHERE id = ?", (account_id,))
            record = cursor.fetchone()

            if not record:
                return jsonify({"code": 404, "msg": "account not found", "data": None}), 404

            record = dict(record)

            # 删除关联的 Cookie 文件
            file_path = record.get("filePath")
            if file_path:
                from pathlib import Path

                from src.core.config import COOKIES_DIR

                cookie_file = COOKIES_DIR / file_path
                if cookie_file.exists():
                    try:
                        cookie_file.unlink()
                        print(f"✅ Cookie 文件已删除: {cookie_file}")
                    except Exception as file_error:
                        print(f"⚠️ 删除 Cookie 文件失败: {file_error}")
                        # 文件删除失败不影响账号删除
                else:
                    print(f"ℹ️ Cookie 文件不存在: {cookie_file}")

            # 删除数据库记录
            cursor.execute("DELETE FROM user_info WHERE id = ?", (account_id,))
            conn.commit()

        return jsonify({"code": 200, "msg": "account deleted successfully", "data": None}), 200

    except Exception as e:
        return jsonify({"code": 500, "msg": str("delete failed!"), "data": None}), 500


@bp.route("/updateUserinfo", methods=["POST"])
def updateUserinfo():
    # 获取JSON数据
    data = request.get_json()

    # 从JSON数据中提取 type 和 userName
    user_id = data.get("id")
    type = data.get("type")
    userName = data.get("userName")
    try:
        # 获取数据库连接
        with sqlite3.connect(db_manager.get_db_path()) as conn:
            conn.row_factory = sqlite3.Row
            cursor = conn.cursor()

            # 更新数据库记录
            cursor.execute(
                """
                           UPDATE user_info
                           SET type     = ?,
                               userName = ?
                           WHERE id = ?;
                           """,
                (type, userName, user_id),
            )
            conn.commit()

        return jsonify({"code": 200, "msg": "account update successfully", "data": None}), 200

    except Exception as e:
        return jsonify({"code": 500, "msg": str("update failed!"), "data": None}), 500

File: apps/backend-node/src/routes/account.ts
/**
 * Account route for omni-post backend (Node.js).
 * Mirrors: apps/backend/src/routes/account.py
 */

import { Router, type Request, type Response } from 'express';
import fs from 'fs';
import { COOKIES_DIR } from '../core/config.js';
import { getPlatformName } from '../core/constants.js';
import { dbManager } from '../db/database.js';
import { getCookieService } from '../services/cookie-service.js';
import { safeJoin } from '../utils/path.js';
import { sendError, sendSuccess } from '../utils/response.js';

export const router = Router();

const VALIDATION_COOLDOWN_MS = 3 * 60 * 60 * 1000; // 3 hours

interface AccountRecord {
  id: number;
  type: number;
  filePath: string;
  userName: string;
  status: number;
  group_id: number | null;
  created_at: string;
  last_validated_at: string | null;
}

/**
 * Perform real-time validation if cooldown expired or forced.
 */
async function validateAccountIfNeeded(account: AccountRecord, force: boolean = false): Promise<AccountRecord> {
    const now = Date.now();
    // Append Z if missing to ensure UTC parsing
    const dateStr = account.last_validated_at && !account.last_validated_at.endsWith('Z') 
        ? account.last_validated_at.replace(' ', 'T') + 'Z' 
        : account.last_validated_at;
    const lastValidated = dateStr ? new Date(dateStr).getTime() : 0;
    
    if (force || (now - lastValidated > VALIDATION_COOLDOWN_MS)) {
        try {
            const isValid = await getCookieService().checkCookie(account.type, account.filePath);
            const status = isValid ? 1 : 0;
            const validatedAt = new Date().toISOString().replace('T', ' ').substring(0, 19);
            
            const db = dbManager.getDb();
            db.prepare('UPDATE user_info SET status = ?, last_validated_at = ? WHERE id = ?')
                .run(status, validatedAt, account.id);
            
            return { ...account, status, last_validated_at: validatedAt };
        } catch (error) {
            console.error(`[Account] Validation failed for ${account.userName}:`, error);
        }
    }
    return account;
}

/**
 * GET /getAccounts
 * Get all accounts (without cookie validation, fast).
 */
router.get('/getAccounts', (_req: Request, res: Response) => {
    try {
        const db = dbManager.getDb();
        const accounts = db.prepare('SELECT * FROM user_info ORDER BY id DESC').all() as AccountRecord[];
        // Convert objects to arrays to match Python's cursor.fetchall() -> list(row) behavior
        const rowsList = accounts.map(account => [
            account.id,
            account.type,
            account.filePath,
            account.userName,
            account.status,
            account.group_id,
            account.created_at,
            account.last_validated_at
        ]);
        sendSuccess(res, rowsList, '获取成功');
    } catch (error: unknown) {
        const err = error as Error;
        sendError(res, 500, `获取账号列表失败: ${err.message}`);
    }
});

/**
 * GET /getValidAccounts
 * Get accounts with cookie validation.
 * Optional: ?id=<accountId> to filter single account.
 * Optional: ?force=true to skip cooldown.
 */
router.get('/getValidAccounts', async (req: Request, res: Response) => {
    try {
        const db = dbManager.getDb();
        const accountId = req.query.id as string | undefined;
        const force = req.query.force === 'true';

        let accounts: AccountRecord[];
        if (accountId) {
            accounts = db.prepare('SELECT * FROM user_info WHERE id = ?').all(Number(accountId)) as AccountRecord[];
        } else {
            accounts = db.prepare('SELECT * FROM user_info ORDER BY id DESC').all() as AccountRecord[];
        }

        // Perform real-time validation for each account
        const validatedAccounts = await Promise.all(
            accounts.map(acc => validateAccountIfNeeded(acc, force))
        );

        const rowsList = validatedAccounts.map(account => [
            account.id,
            account.type,
            account.filePath,
            account.userName,
            account.status,
            account.group_id,
            account.created_at,
            account.last_validated_at,
        ]);

        sendSuccess(res, rowsList, '获取成功');
    } catch (error: any) {
        sendError(res, 500, `获取有效账号失败: ${error.message}`);
    }
});

/**
 * GET /getAccountStatus
 * Get single account status.
 * Optional: ?force=true to skip cooldown.
 */
router.get('/getAccountStatus', async (req: Request, res: Response) => {
    try {
        const db = dbManager.getDb();
        const id = req.query.id as string;
        const force = req.query.force === 'true';

        if (!id) {
            sendError(res, 400, '缺少账号ID');
            return;
        }

        let account = db.prepare('SELECT * FROM user_info WHERE id = ?').get(Number(id)) as any;
        if (!account) {
            sendError(res, 404, '账号不存在');
            return;
        }

        account = await validateAccountIfNeeded(account, force);

        sendSuccess(res, {
            id: account.id,
            type: account.type,
            filePath: account.filePath,
            userName: account.userName,
            status: account.status,
            group_id: account.group_id,
            statusText: account.status === 1 ? '正常' : '异常',
            isValid: account.status === 1,
            platformName: getPlatformName(account.type),
            last_validated_at: account.last_validated_at,
        }, '获取成功');
    } catch (error: any) {
        sendError(res, 500, `获取账号状态失败: ${error.message}`);
    }
});

/**
 * GET /deleteAccount
 * Delete an account.
 */
router.get('/deleteAccount', (req: Request, res: Response) => {
    try {
        const db = dbManager.getDb();
        const id = req.query.id as string;

        if (!id) {
            sendError(res, 400, '缺少账号ID');
            return;
        }

        // Query the record first to get filePath for cookie cleanup (matching Python)
        const record = db.prepare('SELECT * FROM user_info WHERE id = ?').get(Number(id)) as any;
        if (!record) {
            sendError(res, 404, 'account not found');
            return;
        }

        // Delete associated Cookie file (matching Python's account.py behavior)
        if (record.filePath) {
            try {
                const cookieFile = safeJoin(COOKIES_DIR, record.filePath);
                if (fs.existsSync(cookieFile)) {
                    fs.unlinkSync(cookieFile);
                }
            } catch (fileError) {
                console.error(`⚠️ 删除 Cookie 文件失败: ${fileError}`);
            }
        }

        // Delete database record
        db.prepare('DELETE FROM user_info WHERE id = ?').run(Number(id));

        sendSuccess(res, null, 'account deleted successfully');
    } catch (error: any) {
        sendError(res, 500, `删除账号失败: ${error.message}`);
    }
});

/**
 * POST /updateUserinfo
 * Update account information.
 */
router.post('/updateUserinfo', (req: Request, res: Response) => {
    try {
        const db = dbManager.getDb();
        const { id, type, filePath, userName, group_id } = req.body;

        if (!id) {
            sendError(res, 400, '缺少账号ID');
            return;
        }

        const updates: string[] = [];
        const values: any[] = [];

        if (type !== undefined) {
            updates.push('type = ?');
            values.push(type);
        }
        if (filePath !== undefined) {
            updates.push('filePath = ?');
            values.push(filePath);
        }
        if (userName !== undefined) {
            updates.push('userName = ?');
            values.push(userName);
        }
        if (group_id !== undefined) {
            updates.push('group_id = ?');
            values.push(group_id === 0 ? null : group_id);
        }

        if (updates.length === 0) {
            sendError(res, 400, '没有需要更新的字段');
            return;
        }

        values.push(Number(id));
        db.prepare(`UPDATE user_info SET ${updates.join(', ')} WHERE id = ?`).run(...values);

        sendSuccess(res, null, '更新成功');
    } catch (error: any) {
        sendError(res, 500, `更新账号失败: ${error.message}`);
    }
});
